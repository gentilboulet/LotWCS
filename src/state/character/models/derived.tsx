import { ICharacterState } from "./type";

export function chiAura(state: ICharacterState): number {
  if (state.rank) {
    return state.rank;
  }
  return 0;
}

export function chiResplenish(state: ICharacterState): number {
  if (state.rank) {
    return state.rank;
  }
  return 0;
}

export function lake(state: ICharacterState): number {
  if (state.rank) {
    return state.rank + 5;
  }
  return 5;
}

export function river(state: ICharacterState): number {
  if (state.rank) {
    return Math.max(1, state.rank);
  }
  return 1;
}

export function joss(state: ICharacterState): number {
  if (state.rank) {
    return Math.max(1, state.rank);
  }
  return 1;
}

export function maxSkillBonus(state: ICharacterState): number {
  if (state.rank) {
    return Math.max(5, 5 * state.rank);
  }
  return 5;
}
