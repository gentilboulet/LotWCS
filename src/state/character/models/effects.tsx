import * as constants from "../perks/constants/effects";

export type IEffectStatistic =
  | "speed"
  | "footwork"
  | "strike"
  | "damage"
  | "block"
  | "toughness";

export interface IEffectCombatStatistic {
  type: constants.EFFECT_COMBAT_STATISTIC;
  increase: number;
  statistic: IEffectStatistic;
}

export interface IEffectConditionalText {
  type: constants.EFFECT_CONDITIONAL_TEXT;
  text: string[];
}

export interface IEffectChiThresholdIncreaseBaseChi {
  type: constants.EFFECT_CHI_THRESHOLD_INCREASE_BASE_CHI;
  chiIncrease: number;
}

export type IEffect =
  | IEffectCombatStatistic
  | IEffectConditionalText
  | IEffectChiThresholdIncreaseBaseChi;

export function effectToString(effect: IEffect): string {
  switch (effect.type) {
    case constants.EFFECT_COMBAT_STATISTIC:
      return (
        (effect.increase > 0 ? "+" : "") +
        effect.increase +
        " to " +
        effect.statistic
      );
    case constants.EFFECT_CONDITIONAL_TEXT:
      return effect.text.toString();
    case constants.EFFECT_CHI_THRESHOLD_INCREASE_BASE_CHI:
      return (
        (effect.chiIncrease > 0 ? "+" : "") +
        effect.chiIncrease +
        " to Chi Threshold"
      );
  }
}
