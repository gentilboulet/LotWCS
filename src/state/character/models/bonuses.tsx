import { ICharacterState } from "./type";

import { IBonus, isBonus } from "../../../perks/bonuses";
import * as constants from "../../../perks/constants/bonuses";

import * as chi from "./chi";

import { maxSkillBonus } from "./derived";
import * as skills from "./skills";

import * as loresheets from "./loresheets";

export type TBonusesState = IBonus[];

export function createState(): TBonusesState {
  return [];
}

export function applyBonuses(
  draftState: ICharacterState,
  bonuses: IBonus[]
): void {
  bonuses
    .filter((bonus: IBonus) => isBonus(bonus))
    .forEach((bonus: IBonus) => {
      switch (bonus.type) {
        case constants.BONUS_DESTINY:
          draftState.destiny += bonus.value;
          break;
        case constants.BONUS_ENTANGLEMENT:
          draftState.entanglement += bonus.value;
          break;
        case constants.BONUS_CHI:
          chi.increase(draftState.chi, bonus.chi, bonus.value);
          break;
        case constants.BONUS_CULTIVATION:
          chi.increaseCultivation(draftState.chi, bonus.chi, bonus.value);
          break;
        case constants.BONUS_SKILL_RANK:
          const max = maxSkillBonus(draftState);
          // TODO : overflows
          skills.increase(draftState.skills, bonus.skill, max);
          break;
        case constants.BONUS_SPECIALITY:
          // TODO : already speciality
          skills.addSpeciality(
            draftState.skills,
            bonus.skill,
            bonus.speciality
          );
          break;
        case constants.BONUS_LORESHEET:
          loresheets.openLoresheet(draftState.loresheets, bonus.lsUid);
          break;
        case constants.BONUS_LORESHEET_OPTION:
          loresheets.buyLoresheetOption(
            draftState.loresheets,
            bonus.lsUid,
            bonus.uid,
            bonus.payload
          );
          break;
        default:
          break;
      }
    });
}
