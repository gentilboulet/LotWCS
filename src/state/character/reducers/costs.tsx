import { produce, Draft } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/costs";
import { applyCost } from "../models/costs";
import { ICharacterState } from "../models/type";

export type ICostAction = ActionType<typeof actions>;

export const costReducer = produce(
  (draft: Draft<ICharacterState>, action: ICostAction) => {
    switch (action.type) {
      case getType(actions.payCost):
        applyCost(draft, action.payload.cost);
        break;
      default:
    }
  },
);
