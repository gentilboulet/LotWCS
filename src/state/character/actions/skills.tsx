import { createAction } from "typesafe-actions";

import { TSkillName } from "../../../data/skills";
import { ICost } from "../models/costs";

export const skillsBuy = createAction(
  "skills/BUY",
  (name: TSkillName, cost: ICost) => ({ cost, name }),
)();

export const skillSpecialityBuy = createAction(
  "skills/BUY_SPECIALITY",
  (skill: TSkillName, speciality: string, cost: ICost) => {
    if (speciality.length === 0) {
      throw new Error("Empty speciality");
    }
    return { cost, skill, speciality };
  },
)();
