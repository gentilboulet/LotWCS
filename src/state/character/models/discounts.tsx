import { ICost } from "./costs";
import { IStoreState } from "./type";

import { IDiscount, isDiscount } from "../perks/discounts";

export type TDiscountsState = IDiscount[];

export function createState(): TDiscountsState {
  return [];
}

export function pushToDiscounts(
  state: IStoreState,
  discounts: IDiscount[]
): void {
  discounts
    .filter((r: IDiscount) => isDiscount(r))
    .forEach((r: IDiscount) => {
      state.discounts.push(r);
    });
}

export function getDiscountIndexes(
  state: IStoreState,
  predicate: (d: IDiscount) => boolean
) {
  const discountIdx = state.discounts
    .map((discount: IDiscount, idx: number) => {
      if (predicate(discount)) {
        return idx;
      }
      return NaN;
    })
    .filter((value: number) => !isNaN(value));
  if (discountIdx) {
    return discountIdx;
  } else {
    return [];
  }
}

export function updateDiscounts(state: IStoreState, cost: ICost): void {
  // Update the discount
  if (!cost.discounts) {
    return;
  }

  cost.discounts.forEach(costDiscount => {
    state.discounts[costDiscount.idx].value = costDiscount.newValue;

    if (costDiscount.newValue === 0) {
      // Remove the empty discount
      state.discounts.splice(costDiscount.idx, 1);
    }
  });
}
