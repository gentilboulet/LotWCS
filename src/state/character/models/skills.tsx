import { skills as data, TSkillName } from "../../../data/skills";
import { maxSkillBonus } from "./derived";
import { ICharacterState } from "./type";

interface ISkill {
  name: TSkillName;
  specialities: string[];
  value: number;
}

export type TSkillsState = ISkill[];

export function createState(): TSkillsState {
  return Object.keys(data).map((name: TSkillName) => {
    return { name, specialities: [], value: 0 };
  });
}

export function increase(state: TSkillsState, skillName: TSkillName): void {
  getSkill(state, skillName).value += 5;
}

export function isSpecialityPresent(
  state: TSkillsState,
  skillName: TSkillName,
  specialityName: string,
): boolean {
  const specialityIndex = (state.find(
    s => s.name === skillName,
  ) as ISkill).specialities.findIndex(
    (speciality: string) => speciality === specialityName,
  );
  return specialityIndex !== -1;
}

export function addSpeciality(
  state: TSkillsState,
  skillName: TSkillName,
  specialityName: string,
): void {
  if (isSpecialityPresent(state, skillName, specialityName)) {
    throw new Error(
      'Something went wrong, speciality "' +
        specialityName +
        '" for skill "' +
        skillName +
        '" already bought',
    );
  }
  // getSkill(state, skillName).specialities.push(specialityName);
  const spe = getSkill(state, skillName);
  spe.specialities.push(specialityName);
  if (typeof specialityName != "string")
    throw new Error("Only stringm not" + specialityName);
}

export function canBuySkill(
  state: ICharacterState,
  skillName: TSkillName,
): boolean {
  const value = getSkill(state.skills, skillName).value;
  const max = maxSkillBonus(state);
  return !(value + 5 > max);
}

export function canBuySpeciality(
  state: ICharacterState,
  skillName: TSkillName,
  speciality: string,
): boolean {
  return !isSpecialityPresent(state.skills, skillName, speciality);
}

export function getSkill(state: TSkillsState, skillName: TSkillName) {
  console.log("searching for", skillName);
  return state.find(s => s.name === skillName) as ISkill;
}
