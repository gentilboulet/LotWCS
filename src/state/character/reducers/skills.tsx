import produce from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/skills";
import { addSpeciality, increase, TSkillsState } from "../models/skills";

export function skillsReducer(
  state: TSkillsState,
  action: ActionType<typeof actions>,
): TSkillsState {
  return produce(state, draft => {
    switch (action.type) {
      case getType(actions.skillsBuy):
        increase(draft, action.payload.name);
        break;
      case getType(actions.skillSpecialityBuy):
        addSpeciality(draft, action.payload.skill, action.payload.speciality);
        break;
      default:
    }
  });
}
