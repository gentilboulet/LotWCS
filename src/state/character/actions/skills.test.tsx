import { zeroCost } from "../models/costs";
import * as actions from "./skills";
describe("Testing for skill action creators", () => {
  it("should create a buy skill action", () => {
    const skill = "Awareness";
    const a = actions.skillsBuy(skill, zeroCost);
    expect(a.payload.name).toBe(skill);
  });

  it("should create a buy speciality action", () => {
    const skill = "Awareness";
    const speciality = "Hear";
    const a = actions.skillSpecialityBuy(skill, speciality, zeroCost);
    expect(a.payload.skill).toBe(skill);
    expect(a.payload.speciality).toBe(speciality);
    expect(a).toMatchSnapshot();
  });

  it("should not create a buy speciality action on an invalid speciality", () => {
    const skill = "Awareness";
    const speciality = "";
    expect(() =>
      actions.skillSpecialityBuy(skill, speciality, zeroCost)
    ).toThrow();
  });
});
