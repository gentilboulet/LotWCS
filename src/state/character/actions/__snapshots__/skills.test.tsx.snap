// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing for skill action creators should create a buy speciality action 1`] = `
Object {
  "payload": Object {
    "cost": Object {
      "canPay": true,
      "destiny": 0,
      "original": 0,
    },
    "skill": "Awareness",
    "speciality": "Hear",
  },
  "type": "skills/BUY_SPECIALITY",
}
`;
